File: 백준/Silver/2644. 촌수계산/촌수계산.js
const input = require("fs")
  .readFileSync("/dev/stdin", "utf8")
  .trim()
  .split("\n");

// 촌수를 계산해야된다
// 첫번째줄에 사람의 인원수 두번째줄에 관계 촌수를 구해야할 사람 번호
// 세번째줄에 관계수 그다음부터 부모-자식 번호가 주어진다
// bfs로 visited로 마킹하면서 몇번째 촌수인지 하면된다
// 만약 친척 관계가 아니면 -1 출력
const solution = (input) => {
  const peopleN = Number(input[0]);
  const [start, end] = input[1].split(" ").map((v) => Number(v));
  const lineN = Number(input[2]);
  const grid = Array(peopleN + 1)
    .fill(0)
    .map((v) => []);

  for (let i = 3; i < 3 + lineN; i++) {
    const [n1, n2] = input[i].split(" ").map((v) => Number(v));
    grid[n1].push(n2);
    grid[n2].push(n1);
  }

  const visit = Array(peopleN + 1).fill(false);
  visit[start] = true;
  const queue = [[start, 0]];
  let answer = -1;

  while (queue.length) {
    const [curPerson, dist] = queue.shift();
    if (end === curPerson) {
      answer = dist;
      break;
    }

    for (const nextPerson of grid[curPerson]) {
      if (visit[nextPerson]) continue;
      visit[nextPerson] = true;
      queue.push([nextPerson, dist + 1]);
    }
  }

  console.log(answer);
};

solution(input);

File: 백준/Silver/2644. 촌수계산/촌수계산.py
from collections import deque

peopleN = int(input())
start, end = list(map(int, input().split()))
lineN = int(input())
graph = [[] for _ in range(peopleN+1)]

for i in range(lineN):
  x, y = list(map(int, input().split()))
  graph[x].append(y)
  graph[y].append(x)

visited = [False]*(peopleN+1)
visited[start] = True
queue = deque()
queue.append([start, 0])

answer = -1

while len(queue):
  curPerson, dist = queue.popleft()

  if curPerson == end:
    answer = dist
    break

  for nextPerson in graph[curPerson]:
    if not visited[nextPerson]:
      visited[nextPerson] = True
      queue.append([nextPerson, dist+1])

print(answer)